### DISCUSSION TITLE
Showcase: Prediction and Rollback Corrections

### DISCUSSION THREAD

**Discussion by TIHan** [jrouwe/JoltPhysics#1057] [Show and tell]
I wanted to showcase a side project of mine and Jolt's capability of being viable for prediction and rollback corrections.

Related: [Efficient rollbacks to support predict-rollback netcode](https://github.com/jrouwe/JoltPhysics/discussions/1034) - great discussion.
Related: [How is Jolt with networked games?](https://github.com/jrouwe/JoltPhysics/discussions/768)

Video Details:
- Listen Server is top left.
- Client is bottom right.
- Both server and client are running in Debug mode.
- Latency is simulated at **300ms**.
- Green boxes represent a player. *Note: The green box on the client is jittery locally due a bug on my side, not the main part of the showcase though.*
- There are two purple boxes that a client is moving around. Notice on the client-side that interactions with the boxes are instant due to prediction. Any mis-predictions are corrected with rollbacks.

https://github.com/jrouwe/JoltPhysics/assets/1278959/172097ca-d14f-4d91-9f94-8cbb2aef9f10

Implementation details:
- Networking is UDP with snapshots and delta-compression.
- Not using Jolt's built-in Save/Restore state functions. Doing this manually by setting positions/rotations/velocity and ensuring BodyIDs are identical on both client/server.
- Players are using the VirtualCharacter controller.
- This is using .NET 8 with interop calls to Jolt and other native libs.
- Rendering is Vulkan.
- UI is [ImGui](https://github.com/ocornut/imgui).

**Comment by MrOnlineCoder** [comment]
This looks really nice, especially the fact that you are using .NET! Any chance we will be able to look through the source code in near future? ;)

**Comment by cuppajoeman** [comment]
> I wanted to showcase a side project of mine and Jolt's capability of being viable for prediction and rollback corrections.
> 
> Related: [Efficient rollbacks to support predict-rollback netcode](https://github.com/jrouwe/JoltPhysics/discussions/1034) - great discussion. Related: [How is Jolt with networked games?](https://github.com/jrouwe/JoltPhysics/discussions/768)
> 
> Video Details:
> 
>     * Listen Server is top left.
> 
>     * Client is bottom right.
> 
>     * Both server and client are running in Debug mode.
> 
>     * Latency is simulated at **300ms**.
> 
>     * Green boxes represent a player. _Note: The green box on the client is jittery locally due a bug on my side, not the main part of the showcase though._
> 
>     * There are two purple boxes that a client is moving around. Notice on the client-side that interactions with the boxes are instant due to prediction. Any mis-predictions are corrected with rollbacks.
> 
> 
> jolt_prediction_rollback_corrections.mp4
> 
> Implementation details:
> 
>     * Networking is UDP with snapshots and delta-compression.
> 
>     * Not using Jolt's built-in Save/Restore state functions. Doing this manually by setting positions/rotations/velocity and ensuring BodyIDs are identical on both client/server.
> 
>     * Players are using the VirtualCharacter controller.
> 
>     * This is using .NET 8 with interop calls to Jolt and other native libs.
> 
>     * Rendering is Vulkan.
> 
>     * UI is [ImGui](https://github.com/ocornut/imgui).

Just saw this today, I'm literally working on the same thing in c++ and opengl would love to see the source code for this!

Also I want to double check, you're running an instance of jolt on the client and the server, then whenever your game updates come through onto the client you simply iterate through all your physics objects and manually set their properties?

**Reply by captainkidd5** [reply]
would also like to know the answer to your last question here

**Reply by TIHan** [reply]
Apologies for not getting back to this.

> Any chance we will be able to look through the source code in near future?

> would love to see the source code for this!

I made it open source today and upgraded to Jolt 5.2.0: https://github.com/TIHan/Evergreen 
It's just a side-project to learn how game engines work and using my own .NET language, [Oly](https://github.com/TIHan/Oly), to do it.

> Also I want to double check, you're running an instance of jolt on the client and the server, then whenever your game updates come through onto the client you simply iterate through all your physics objects and manually set their properties?

Yes, this is correct. The function that does this is [here](https://github.com/TIHan/Evergreen/blob/cf7ae3f0093343465b1dcaef1c6bb4d6f9beab98/src/managed/Game/Client/Systems/PhysicsSystem.oly#L182). It's called after we update the client's world from the server [here](https://github.com/TIHan/Evergreen/blob/cf7ae3f0093343465b1dcaef1c6bb4d6f9beab98/src/managed/Game/Client/ClientScene.oly#L136) , then the rollback corrections will be called right after as you can see:
```csharp
            clientWorld.SetPhysicsState()

            // Because rollback corrections are expensive, skip it when catching-up.
            if (!isCatchingUp)
                clientWorld.RollbackCorrections(time, this.clientEntityId)

            clientWorld.UpdateTransforms()
```

Other physics relevant files:
- [Native C++ backend using Jolt](https://github.com/TIHan/Evergreen/blob/cf7ae3f0093343465b1dcaef1c6bb4d6f9beab98/src/native/libs/Evergreen.Physics/egJolt.cpp)
- [Higher-level API using the backend](https://github.com/TIHan/Evergreen/blob/cf7ae3f0093343465b1dcaef1c6bb4d6f9beab98/src/managed/Physics/Evergreen.Physics.olyx)
- [Shared physics system](https://github.com/TIHan/Evergreen/blob/cf7ae3f0093343465b1dcaef1c6bb4d6f9beab98/src/managed/Game/Shared/Systems/PhysicsSystem.oly)
- [Server physics system](https://github.com/TIHan/Evergreen/blob/cf7ae3f0093343465b1dcaef1c6bb4d6f9beab98/src/managed/Game/Server/Systems/PhysicsSystem.oly)
- [Client physics system](https://github.com/TIHan/Evergreen/blob/cf7ae3f0093343465b1dcaef1c6bb4d6f9beab98/src/managed/Game/Client/Systems/PhysicsSystem.oly)
    - [Rollback corrections implementation](https://github.com/TIHan/Evergreen/blob/cf7ae3f0093343465b1dcaef1c6bb4d6f9beab98/src/managed/Game/Client/Systems/PhysicsSystem.oly#L241)
    
cc @cuppajoeman @captainkidd5 @MrOnlineCoder 

---

I hope this helps! I know this isn't typical given it's in a language used by a single person.
